### ISSUE TITLE
Confusing error messages

### ISSUE THREAD

**Issue by paulmillar** [italiangrid/voms-api-java#18]
I recently changed dCache to show AC validation errors but unfortunately the error messages are not terribly helpful.  Here's the output I get from concatenating `getMessage()` for each error:

    [CAnL certificate validation error: No trusted CA certificate was found for the certificate chain, CAnL certificate validation error: Trusted issuer of this certificate was not established, LSC validation failed: AA certificate chain embedded in the VOMS AC failed certificate validation!, AC signature verification failure: no valid VOMS server credential found.]

There's three problems here: too much text, not enough detail, and difficult-to-understand causality.  The result is that it took quite some time to understand the phrasing and I still don't know what I should do to fix it.

For an example of too much text: the two errors:

    CAnL certificate validation error: No trusted CA certificate was found for the certificate chain, CAnL certificate validation error: Trusted issuer of this certificate was not established

Apart from the "CAnL certificate validation error" being redundant, I believe this is really the same error, and could be summarised by:

    Cannot establish chain-of-trust for this certificate.

An example of not enough detail is "LSC validation failed: AA certificate chain embedded in the VOMS AC failed certificate validation!".  It doesn't say why!

The third problem is more an architectural one: the error reporting is a simple list of problems.  The causal relationship in the error messages is unclear.

I think the above message could be rewritten as:

    Cannot establish chain-of-trust for this certificate: AA certificate failed certificate validation via LSC and no explicit trust of the AA certificate.

In other words, the first two errors (which are really one error) are the result of the combined failed attempt at using LSC  and the failure to find the VOMS server certificate in the trust-store.

This isn't clear from a list of errors.

**Comment by andreaceccanti** [comment]
VOMS forwards CANL errors to the caller. The whole error here (AFAIU) is that LSC validation failed because your trust store was missing the CA certificate that signed the VOMS AA certificate embedded in the AC.

I agree that the first two CANL error messages are redundant, I think I've also opened an issue for
this time ago.

The output you show is a partial view of the full error message. If you would have printed all the information, you would have seen which was the certificate for which no CA was found.

VOMS error message are the following:

1. LSC validation failed: AA certificate chain embedded in the VOMS AC failed certificate validation!
2. AC signature verification failure: no valid VOMS server credential found.

The AA certificate chain validation can fail for many reasons: this is why we forward the CANL error
message for that, so that the called has all the information to understand what's happening.

The second tells you that the AC signature verification failed since the API could not find a valid certificate to check the signature.

I think the current list of errors provides all the info to understand what's happening, while the single error message you propose is much more confusing.

**Comment by andreaceccanti** [comment]
Regarding causality: in this case the causality is linear, the first error causes the second etc.
I will have to check if there are cases when this is *not* true.
I agree that in such case having information that explicitly correlates one error to the other may be needed.

**Comment by paulmillar** [comment]
Hi Andrea,

Thanks for the quick reply.

WRT this specific error -- the odd thing is that I thought I had included the corresponding CA in /etc/grid-security/certificates, but apparently this isn't being accepted.  So I'm kinda stuck with insufficient information.

Would it be possible to remove the colons from the messages, and perhaps make the messages shorter?

I'd like to concat the messages with colons, as this is how we represent problems elsewhere in dCache.  However, if the messages contain colons then this isn't possible.

Cheers,
Paul.

**Comment by andreaceccanti** [comment]
Hi Paul,

> the odd thing is that I thought I had included the corresponding CA in /etc/grid-security/certificates, but 
> apparently this isn't being accepted.

That depends on the CANL validator. How are you creating the validator?  

> Would it be possible to remove the colons from the messages, and perhaps make the messages shorter?

I fail to see the reason. 

1. I don't think error message are too long. I think that actually the messages are quite clear and pretty informative (I am speaking of the VOMS ones, not the CANL's)
2. Validation error message come with a validation error code, which is an enum, and from this enum you can derive whatever message you see more appropriate for dCache

https://github.com/italiangrid/voms-api-java/blob/master/src/main/java/org/italiangrid/voms/error/VOMSValidationErrorMessage.java#L34

**Comment by andreaceccanti** [comment]
Hi Paul,

Did the above suggestion help in moving forward? Can I close this issue?

**Comment by paulmillar** [comment]
Hi Andrea,

Sorry, I haven't been able to resolve the issue.

I have (what is claimed to be) the CA that signed the VOMS certificate in /etc/grid-security/certificates, along with signing_policy and namespaces files.

You can see the code that instantiates the validator here:

https://github.com/dCache/dcache/blob/master/modules/gplazma2-voms/src/main/java/org/dcache/gplazma/plugins/VomsPlugin.java

I don't know what's wrong and unfortunately the error messages haven't clarified the situation.

Cheers,

Paul.

**Comment by andreaceccanti** [comment]
Hi Paul,

1. the cerificate validator you create does not refresh information from
the trust anchors dir.
is that what you want? To enable refresh you have to specify a
trustAnchorsRefreshInterval
to the builder (the default is 0, i.e. no refresh).
2. you don't register a listener for other errors that may be raised by the
CANL validator
when loading/refreshing trust anchors. I would recommend you do that.

This is how we initialise the CANL validator in StoRM webdav, probably it
can help in getting
more feedback on why things are failing:

https://github.com/italiangrid/storm-webdav/blob/master/src/main/java/org/italiangrid/storm/webdav/spring/AppConfig.java#L115

Let me know if this helps.

Cheers,
Andrea

2016-01-25 9:47 GMT+01:00 Paul Millar <notifications@github.com>:

**Comment by paulmillar** [comment]
Thanks for the refresh information tip -- that certainly looks like a bug.  I'll fix that next.

I didn't register a listener because I thought `VOMSACValidator#validateWithResult` would give me all necessary information as the list of `VOMSValidationResult`.

Just to be sure, I added a `StoreUpdateListener` and `ValidationErrorListener` and attempted to use the problematic proxy.  The StoreUpdateListener didn't fire and the ValidationErrorLister fired twice:

    25 Jan 2016 11:09:32 (gPlazma) [qxM:1:srm2:ls SRM-prometheus Login AUTH voms] Validation error: error affecting the whole chain (category: X509_CHAIN): No trusted CA certificate was found for the certificate chain

    25 Jan 2016 11:09:32 (gPlazma) [qxM:1:srm2:ls SRM-prometheus Login AUTH voms] Validation error: error at position 0 in chain, problematic certificate subject: CN=snf-683051.vm.okeanos.grnet.gr,OU=PDP,O=Example,C=NL (category: X509_CHAIN): Trusted issuer of this certificate was not established


    paul@celebrimbor:~/git/dCache$ voms-proxy-info -chain -all
    === Proxy Chain Information ===
    subject   : /O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl
    issuer    : /O=Grid/OU=GlobusTest/CN=Globus Simple CA for Demo Portal
    key usage : Digital Signature, Key Encipherment, Data Encipherment
    strength  : 2048 bits
    timeleft  : 8358:08:23
    
    === Proxy Information ===
    subject   : /O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl/CN=1095114846
    issuer    : /O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl
    identity  : /O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl
    type      : RFC compliant proxy
    strength  : 2048 bits
    path      : /tmp/x509up_u1000
    timeleft  : 11:47:38
    key usage : Digital Signature, Key Encipherment, Data Encipherment
    === VO testvo extension information ===
    VO        : testvo
    subject   : /O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl
    issuer    : /C=NL/O=Example/OU=PDP/CN=snf-683051.vm.okeanos.grnet.gr
    attribute : /testvo/Role=NULL/Capability=NULL
    timeleft  : 11:47:38
    uri       : snf-683051.vm.okeanos.grnet.gr:15000
    paul@celebrimbor:~/git/dCache$

**Comment by andreaceccanti** [comment]
Hi Paul,

voms-proxy-info uses voms-api as well, so if voms-proxy-info works fine, so should gPlazma.

voms-proxy-info shows info about this proxy:
/O=Grid/OU=GlobusTest/CN=9bd9745d-dc80-44fa-97cb-928d1ef6f5df@surfguest.nl/CN=1095114846

while you get a failure for this cert:
CN=snf-683051.vm.okeanos.grnet.gr,OU=PDP,O=Example,C=NL

so this looks strange.

**Comment by andreaceccanti** [comment]
Is the namespace/signing policy configured correctly for "CN=snf-683051.vm.okeanos.grnet.gr,OU=PDP,O=Example,C=NL" ?

Now I see that that subjects signs a certificate with a completely different subject and if namespaces and signing policies aren't configured to allow this you may incur into problems...

**Comment by paulmillar** [comment]
Hi Andrea,

I agree this is pretty terrible, but at least the CA is signing with a common prefix: both the CA's certificate and the VOMS server's start `/C=NL/O=Example/OU=PDP`.

Here's some details of the deployment.

The `.lsc` file:

```
[root@prometheus ~]# cat /etc/grid-security/vomsdir/testvo/snf-683051.vm.okeanos.grnet.gr.lsc 
/C=NL/O=Example/OU=PDP/CN=snf-683051.vm.okeanos.grnet.gr
/C=NL/O=Example/OU=PDP/CN=rootCA
[root@prometheus ~]# 
```

The CA that signed the VOMS server certificate:

```
[root@prometheus ~]# openssl x509 -in /etc/grid-security/certificates/AARC-pilot-root-CA.pem -noout -subject -hash
subject= /C=NL/O=Example/OU=PDP/CN=rootCA
2a0dd81a
```

Here's the namespace/signing_policy files:

```
[root@prometheus ~]# cat /etc/grid-security/certificates/AARC-pilot-root-CA.signing_policy 
#  Root CA for AARC pilot used to sign VOMS server certificates
#
access_id_CA   X509    '/C=NL/O=Example/OU=PDP/CN=rootCA'
pos_rights     globus  CA:sign
cond_subjects  globus  '"/C=NL/O=Example/OU=PDP/CN=rootCA" "/C=NL/O=Example/OU=PDP/CN=*"'
##############################################################################
[root@prometheus ~]# cat /etc/grid-security/certificates/AARC-pilot-root-CA.namespaces 
##############################################################################
#NAMESPACES-VERSION: 1.0
#
#
TO Issuer "/C=NL/O=Example/OU=PDP/CN=rootCA" \
  PERMIT Subject "/C=NL/O=Example/OU=PDP/.*"

[root@prometheus ~]# 
```

And finally, here are the hash entries as sym-links to the corresponding files:

```
[root@prometheus ~]# ls -l /etc/grid-security/certificates/2a0dd81a*
lrwxrwxrwx 1 root root 22 Jan 12 09:35 /etc/grid-security/certificates/2a0dd81a.0 -> AARC-pilot-root-CA.pem
lrwxrwxrwx 1 root root 29 Jan 12 09:41 /etc/grid-security/certificates/2a0dd81a.namespaces -> AARC-pilot-root-CA.namespaces
lrwxrwxrwx 1 root root 33 Jan 12 09:41 /etc/grid-security/certificates/2a0dd81a.signing_policy -> AARC-pilot-root-CA.signing_policy
[root@prometheus ~]# 
```

I think my point still stands: the VOMS library is refusing to validate the voms proxy (rightly or wrongly, I can't say) but it isn't providing sufficient information to understand why.

**Comment by andreaceccanti** [comment]
Hi Paul,

can you list *all* the files for the AARC-pilot-root-CA in
/etc/grid-security/certificates, including permissions etc.?

> I think my point still stands: the VOMS library is refusing to validate
the voms proxy (rightly or wrongly, I can't say) but it isn't
> providing sufficient information to understand why.

it's CANL, not VOMS,  that refuses to validate your cert AFAIU.

A.


2016-01-26 11:06 GMT+01:00 Paul Millar <notifications@github.com>:

**Comment by paulmillar** [comment]
Hi Andrea,

Here's the output from `ls -l`.  There are two AARC CAs: one is used to generate EECs (AARC-pilot-CA.pem), the other is used to sign the VOMS server's certificate (AARC-pilot-root-CA).

I've included both here:

```
[root@prometheus ~]# ls -l /etc/grid-security/certificates/AARC*
-rw-r--r-- 1 root root  226 Jan 12 09:41 /etc/grid-security/certificates/AARC-pilot-CA.namespaces
-rw-r--r-- 1 root root 1931 Jan 11 11:27 /etc/grid-security/certificates/AARC-pilot-CA.pem
-rw-r--r-- 1 root root  272 Jan 12 09:40 /etc/grid-security/certificates/AARC-pilot-CA.signing_policy
-rw-r--r-- 1 root root  201 Jan 14 11:27 /etc/grid-security/certificates/AARC-pilot-root-CA.namespaces
-rw-r--r-- 1 root root 1192 Jan 12 09:34 /etc/grid-security/certificates/AARC-pilot-root-CA.pem
-rw-r--r-- 1 root root  324 Jan 12 09:40 /etc/grid-security/certificates/AARC-pilot-root-CA.signing_policy
[root@prometheus ~]# 
```

The hashes of the two CAs:
```
[root@prometheus ~]# openssl x509 -in /etc/grid-security/certificates/AARC-pilot-CA.pem -noout -hash
93df451c
[root@prometheus ~]# openssl x509 -in /etc/grid-security/certificates/AARC-pilot-root-CA.pem -noout -hash
2a0dd81a
```

The directory listings of the sym-links:

```
[root@prometheus ~]# ls -l /etc/grid-security/certificates/93df451c*
lrwxrwxrwx 1 root root 17 Jan 11 11:29 /etc/grid-security/certificates/93df451c.0 -> AARC-pilot-CA.pem
lrwxrwxrwx 1 root root 24 Jan 11 11:34 /etc/grid-security/certificates/93df451c.namespaces -> AARC-pilot-CA.namespaces
lrwxrwxrwx 1 root root 28 Jan 11 11:34 /etc/grid-security/certificates/93df451c.signing_policy -> AARC-pilot-CA.signing_policy
[root@prometheus ~]# ls -l /etc/grid-security/certificates/2a0dd81a*
lrwxrwxrwx 1 root root 22 Jan 12 09:35 /etc/grid-security/certificates/2a0dd81a.0 -> AARC-pilot-root-CA.pem
lrwxrwxrwx 1 root root 29 Jan 12 09:41 /etc/grid-security/certificates/2a0dd81a.namespaces -> AARC-pilot-root-CA.namespaces
lrwxrwxrwx 1 root root 33 Jan 12 09:41 /etc/grid-security/certificates/2a0dd81a.signing_policy -> AARC-pilot-root-CA.signing_policy
[root@prometheus ~]# 
```

HTH,

Paul.

**Comment by paulmillar** [comment]
BTW, is there an easy way to extract the VOMS server certificate from the proxy?  I'd like to double-check that it really is signed by the VO certificate AARC claims it is.

**Comment by andreaceccanti** [comment]
Well, the easiest that comes to mind involve setting a breakpoint here:

https://github.com/italiangrid/voms-api-java/blob/master/src/main/java/org/italiangrid/voms/ac/impl/DefaultVOMSValidationStrategy.java#L162

and then here:

https://github.com/italiangrid/voms-api-java/blob/master/src/main/java/org/italiangrid/voms/ac/impl/DefaultVOMSValidationStrategy.java#L356

to have a detailed look into the CANL validation process.

**Comment by paulmillar** [comment]
Unfortunately, the AARC pilot was broken for a few days, but they fixed it today, so I investigated further.

The ultimate problem seems to stem from `eu.emi.security.authn.x509.helpers.pkipath.bc.CertPathValidatorUtilities#findTrustAnchor2` returning `null`.  This is called from `eu.emi.security.authn.x509.helpers.pkipath.NonValidatingCertPathBuilder#build`.

My reading of this situation is that either:
    the certificate their VOMS server is using when generating the AC was not signed by the CA they claim it was,
or:
    I haven't installed the CA certificate correctly in /etc/grid-security/certificates directory.

Does this sound correct?

To try to test this, I've started added some code in dCache to allow extracting of the AC certificate(s).  These are written as PEM-encoded files that start `-----BEGIN ATTRIBUTE CERTIFICATE-----` as that's the somewhat broken code in BouncyCastle does.

I guess I would also need to extract the VOMS AC-generating certificate (which is also in the voms proxy, right?) and then attempt to manually verify the chain.

Unfortunately, openssl doesn't seem to support ACs, so I cannot introspect these ACs.

Any ideas what would be the best approach here?

**Comment by andreaceccanti** [comment]
The easiest way to get the VOMS server certificate is:

openssl s_client -connect <voms_host>:<voms_port> | openssl x509 -text

**Comment by paulmillar** [comment]
Yeah, the problem there is that I don't know (and perhaps can't access) the voms server endpoint.  In the AARC deployment, its a web-portal that contacts the VOMS server and generates the voms-proxy on my behalf.  I'm copying it from the web-page to try to get it to work with dCache.

**Comment by andreaceccanti** [comment]
However you can deduce the endpoint hostname from the LSC file, and assume
they're using the default port for VOMS (15000).
Otherwise, I would ask the web portal developer.
If this is impossible, follow my former advice and set the breakpoints in
the VOMS code so that you can look at the certificate extracted
from the AC.


2016-02-05 14:03 GMT+01:00 Paul Millar <notifications@github.com>:

**Comment by paulmillar** [comment]
Turns out that the client-facing certificate was signed by the expected CA:

    openssl s_client -connect snf-683051.vm.okeanos.grnet.gr:15000 | openssl x509 -outform pem -out /tmp/AARC-voms-server.pem
    paul@celebrimbor:~$ scp /tmp/AARC-voms-server.pem root@prometheus:
    AARC-voms-server.pem                                                                                                 100% 1285     1.3KB/s   00:00    
    paul@celebrimbor:~$ ssh root@prometheus
    Last login: Fri Feb  5 15:46:51 2016 from zitpcx19643.desy.de
    
    Welcome to DESY IT
    
    [root@prometheus ~]# openssl verify -verbose -CAfile /etc/grid-security/certificates/AARC-pilot-root-CA.pem AARC-voms-server.pem
    AARC-voms-server.pem: OK
    [root@prometheus ~]# 

Could the VOMS server be using a different certificate (i.e., not the client-facing certificate) to sign the ACs?

**Comment by andreaceccanti** [comment]
No.

**Comment by andreaceccanti** [comment]
Would you mind showing the output of:
openssl x509 -in VOMSCERT.pem -noout -text

**Comment by paulmillar** [comment]
Found it!

It turns out that the VOMS validation requires the old (MD5) hash entries in the trust store, at least for `.lsc` files.

I've opened a fresh issue describing the problem here: #19.

However, this issue still remains.  I found the problem quite by chance.  Nothing in the error messages tells me what is the problem.

**Comment by andreaceccanti** [comment]
Paul,

it's CANL that does certificate validation, not VOMS.

**Comment by paulmillar** [comment]
Yup, sure looks like a CANL bug.

But, for VOMS validation, dCache isn't using CANL directly, so I think we should keep this issue open and "a voms-api-java developer"  (i.e., you ;-) should open a ticket against CANL.

Does that seem reasonable?

**Comment by andreaceccanti** [comment]
> Does that seem reasonable?
Hm, not sure... 
Actually this is not a bug in CANL :)
I would say there are two bugs:

- One in the way your CA is packaged (your CA wouldn't work on a machine that has openssl < 1.0). All CAs in the IGTF distribution ship with MD5 hashes for the links for exactly this reason.
- One in the VOMS CertificateValidatorBuilder that assumes as a conservative default that md5 hashes are **always** shipped by CA packages and does not expose a way to change its default.

The CertificateValidatorBuilder was built to shield users from the the rather ugly OpensslCertChainValidator chain of constructors:

https://github.com/eu-emi/canl-java/blob/master/src/main/java/eu/emi/security/authn/x509/impl/OpensslCertChainValidator.java#L80

There is a parameter that selects whether the validator should assume an environment where
openssl1 or openssl0.9 is running (basically to change the hash function used to resolve signing policies, namespace etc.). 

It's not possible to tune such parameter with the CertificateValidatorBuilder. So this is something I will fix in VOMS (will open another issue in our JIRA tracker). Regarding the error messages, if you think something was missing (like "signing policy not found for CA certificate X") I think you should
open a bug to CANL ;)

Thanks,
A.

**Comment by andreaceccanti** [comment]
Related issue:
https://issues.infn.it/jira/browse/VOMS-703

**Comment by andreaceccanti** [comment]
btw, as another workaround in your code you can create the OpensslCertChainValidator directly and set the parameter according to your installation:

https://github.com/italiangrid/voms-api-java/blob/master/src/main/java/org/italiangrid/voms/util/CertificateValidatorBuilder.java#L211

**Comment by andreaceccanti** [comment]
Closing this as 3.0.6/3.1.0/ 3.2.0 solve the issue.

**Comment by andreaceccanti** [comment]
http://italiangrid.github.io/voms/release-notes/voms-api-java/3.0.6/
http://italiangrid.github.io/voms/release-notes/voms-api-java/3.1.0/
http://italiangrid.github.io/voms/release-notes/voms-api-java/3.2.0/
